xapian-bindings/xapianletor-headers.i

```diff
@@ -0,0 +1,84 @@
+%{
+/* xapianletor-headers.i: Getting SWIG to parse Xapian's C++ headers.
+ *
+ * Copyright 2004,2006,2011,2012,2013,2014,2015,2016,2017,2019 Olly Betts
+ * Copyright 2014 Assem Chelli
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+%}
+
+/* Ignore anything ending in an underscore, which is for internal use only: */
+%rename("$ignore", regexmatch$name="_$") "";
+
+/* For other languages, SWIG already renames operator() suitably. */
+#if defined SWIGJAVA || defined SWIGPHP || defined SWIGTCL
+%rename(apply) *::operator();
+#elif defined SWIGCSHARP
+%rename(Apply) *::operator();
+#endif
+
+/* Ignore these for all classes: */
+%ignore operator==;
+%ignore operator!=;
+%ignore operator<;
+%ignore operator>;
+%ignore operator<=;
+%ignore operator>=;
+%ignore operator+;
+%ignore difference_type;
+%ignore iterator_category;
+%ignore value_type;
+%ignore max_size;
+%ignore swap;
+%ignore iterator;
+%ignore const_iterator;
+%ignore size_type;
+%ignore unserialise(const char **, const char *);
+%ignore release();
```

Comment:
Everything above this seems common with core, and could perhaps be shared?

Reply:
Probably a good plan, but perhaps better to address after merging rather than dragging out this PR further.

The current structuring is partly a result of the history - originally we had a copy of the C++ API declarations we wanted to wrap here in the bindings, which was a maintenance headache so we switched over to having SWIG parse the C++ API headers instead. 

We switched class by class with the "wrap via API headers" code being added to `xapian-headers.i` and the old declarations living in `xapian.i` (e.g. see 553db57962ba6d2d80e7518ad53441db5dbb2310 for TermIterator being switched).  Once everything was switched, `xapian.i` went away completely.

`xapian-head.i` is mostly stuff that needs including early on, but also has fallback implementations of the remote backend factory functions which perhaps should live somewhere else.  Arguably that should be handled in xapian-core, though if you're disabling backends you don't need to reduce the library size (e.g. for embedded use) then it's nicer not to have stub implementations.  They could be inlined from the core API headers if disabled though, which should mean no overhead in a program which doesn't call them.

Reply:
Yeah, I'd be happy to punt this to the future.